File: reference/PCF-SIG-v1.0/examples/gen_testvector.rs
//! Generates the canonical PCF-SIG v1.0 test-vector file used in spec
//! section 19.
//!
//! Run with: `cargo run --example gen_testvector -- <output-path>`
//! (defaults to ./pcfsig_testvector.bin).
//!
//! The Ed25519 keypair is generated deterministically from a fixed 32-byte
//! seed of 0x00..0x1F, so independent implementations can reproduce the file
//! byte-for-byte.

use std::io::Cursor;

use pcf::{Container, HashAlgo};
use pcf_sig::{sign_partitions, verify_all, DataRecheck, ManifestVerdict, SigningMaterial};
use sha2::{Digest, Sha256};

fn main() {
    let path = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "pcfsig_testvector.bin".to_string());

    let seed: [u8; 32] = std::array::from_fn(|i| i as u8);
    let signer = SigningMaterial::ed25519_from_seed(&seed);

    let mut c = Container::create_with(Cursor::new(Vec::new()), 8, HashAlgo::Sha256).unwrap();

    // Partition "alpha": the partition to be signed.
    c.add_partition(
        0x0000_0010,
        [0x11u8; 16],
        "alpha",
        b"Hello, PCF-SIG!",
        0,
        HashAlgo::Sha256,
    )
    .unwrap();

    // Sign it. This adds a PCFSIG_KEY partition (uid = 0x22..) and a
    // PCFSIG_SIG partition (uid = 0x33..).
    sign_partitions(
        &mut c,
        &signer,
        &[[0x11u8; 16]],
        [0x33u8; 16], // sig partition uid
        [0x22u8; 16], // key partition uid
        0,            // signed_at = unspecified
        "pcfsig",
        "pcfkey",
    )
    .unwrap();

    // Compact to the canonical layout and re-verify.
    let image = c.compacted_image().unwrap();
    std::fs::write(&path, &image).unwrap();

    let mut v = Container::open(Cursor::new(image.clone())).unwrap();
    v.verify().unwrap();
    let reports = verify_all(&mut v, DataRecheck::Recompute).unwrap();
    assert_eq!(reports.len(), 1);
    assert!(matches!(reports[0].verdict, ManifestVerdict::Valid));

    let digest = Sha256::digest(&image);
    let hex: String = digest.iter().map(|b| format!("{b:02x}")).collect();
    eprintln!("wrote {} ({} bytes)", path, image.len());
    eprintln!("sha256 = {hex}");
    eprintln!(
        "signer fingerprint = {}",
        signer
            .fingerprint()
            .iter()
            .map(|b| format!("{b:02x}"))
            .collect::<String>()
    );
}

File: implementations/php/pcf-sig/examples/gen_testvector.php
<?php

declare(strict_types=1);

/**
 * Generates the canonical PCF-SIG v1.0 test-vector file. Run with
 * `php examples/gen_testvector.php <output-path>` (defaults to
 * ./pcfsig_testvector.bin).
 *
 * The Ed25519 keypair is generated deterministically from a fixed 32-byte seed
 * of 0x00..0x1F, so independent implementations can reproduce the file
 * byte-for-byte.
 */

require __DIR__ . '/../vendor/autoload.php';

use Kduma\PCF\Container;
use Kduma\PCF\HashAlgo;
use Kduma\PCF\Storage\MemoryStorage;
use Kduma\PCFSIG\ManifestVerdict;
use Kduma\PCFSIG\SignPartitions;
use Kduma\PCFSIG\SigningMaterial;
use Kduma\PCFSIG\Verify;

$path = $argv[1] ?? 'pcfsig_testvector.bin';

$seed = '';
for ($i = 0; $i < 32; ++$i) {
    $seed .= chr($i);
}
$signer = SigningMaterial::ed25519FromSeed($seed);

$c = Container::createWith(new MemoryStorage(), 8, HashAlgo::Sha256);
$c->addPartition(
    0x10,
    str_repeat("\x11", 16),
    'alpha',
    'Hello, PCF-SIG!',
    0,
    HashAlgo::Sha256,
);
SignPartitions::run(
    $c,
    $signer,
    [str_repeat("\x11", 16)],
    str_repeat("\x33", 16),
    str_repeat("\x22", 16),
    0,
    'pcfsig',
    'pcfkey',
);

$image = $c->compactedImage();
file_put_contents($path, $image);

$verifier = Container::open(new MemoryStorage($image));
$verifier->verify();
$reports = Verify::allWithRecheck($verifier);
if (count($reports) !== 1 || $reports[0]->verdict !== ManifestVerdict::Valid) {
    fwrite(STDERR, "generated vector does not self-verify\n");
    exit(1);
}

fprintf(STDERR, "wrote %s (%d bytes)\n", $path, strlen($image));
fprintf(STDERR, "sha256 = %s\n", bin2hex(hash('sha256', $image, true)));
fprintf(STDERR, "signer fingerprint = %s\n", bin2hex($signer->fingerprint()));
